lib/luahid.c

```diff
@@ -181,63 +184,103 @@ typedef struct {
 
-static void luahid_preprobe(lua_State *L, struct hid_device *hdev, luahid_probe_arg_t *arg)
+static void luahid_preprobe(lua_State *L, struct hid_device *hdev, void *arg)
 {
-	luahid_newtable(L, arg, driver_data);
+	luahid_probe_arg_t *probe_arg = arg;
+	probe_arg->ret = 0;
+	luahid_newtable(L, probe_arg->id, driver_data);
 }
 
-static void luahid_prereport_fixup(lua_State *L, struct hid_device *hdev, luahid_report_fixup_arg_t *arg)
+static void luahid_prereport_fixup(lua_State *L, struct hid_device *hdev, void *arg)
 {
 	luahid_pushhdev(L, hdev);
 	luahid_pushinfo(L, -4, hdev);
 	lunatik_object_t *data = luahid_getdescriptor(L, luahid_gethid(hdev));
+	luahid_report_fixup_arg_t *report_fixup_arg = arg;
 	if (data)
-		luadata_reset(data, arg->rdesc, arg->rsize, LUADATA_OPT_NONE);
+		luadata_reset(data, report_fixup_arg->rdesc, report_fixup_arg->rsize, LUADATA_OPT_NONE);
 }
 
-static void luahid_preraw_event(lua_State *L, struct hid_device *hdev, luahid_raw_event_arg_t *arg)
+static void luahid_preraw_event(lua_State *L, struct hid_device *hdev, void *arg)
 {
 	luahid_pushhdev(L, hdev);
 	luahid_pushinfo(L, -4, hdev);
-	luahid_pushreport(L, arg->report);
+	luahid_raw_event_arg_t *raw_event_arg = arg;
+	raw_event_arg->ret = 0;
+	luahid_pushreport(L, raw_event_arg->report);
 	lunatik_object_t *data = luahid_getdescriptor(L, luahid_gethid(hdev));
 	if (data)
-		luadata_reset(data, arg->data, arg->size, LUADATA_OPT_NONE);
+		luadata_reset(data, raw_event_arg->data, raw_event_arg->size, LUADATA_OPT_NONE);
 }
 
-static void luahid_postprobe(lua_State *L, struct hid_device *hdev, luahid_probe_arg_t *arg)
+static void luahid_postprobe(lua_State *L, struct hid_device *hdev, void *arg)
 {
+	luahid_probe_arg_t *probe_arg = arg;
 	if (lua_istable(L, -1)) {
 		lua_pushlightuserdata(L, hdev);
 		lua_pushvalue(L, -2);
 		lua_settable(L, -4);
 		hid_set_drvdata(hdev, luahid_gethid(hdev));
+		probe_arg->ret = 0;
+	} else if (lua_isinteger(L, -1)) {
+		probe_arg->ret = lua_tointeger(L, -1);
+	} else {
+		probe_arg->ret = 0;
 	}
-	lua_pushinteger(L, 0);
 }
 
-static void luahid_postreport_fixup(lua_State *L, struct hid_device *hdev, luahid_report_fixup_arg_t *arg)
+static void luahid_postreport_fixup(lua_State *L, struct hid_device *hdev, void *arg)
 {
 	lunatik_object_t *data = luahid_getdescriptor(L, luahid_gethid(hdev));
 	if (data)
 		luadata_clear(data);
 }
 
-static void luahid_postraw_event(lua_State *L, struct hid_device *hdev, luahid_raw_event_arg_t *arg)
+static void luahid_postraw_event(lua_State *L, struct hid_device *hdev, void *arg)
 {
+	luahid_raw_event_arg_t *raw_event_arg = arg;
 	lunatik_object_t *data = luahid_getdescriptor(L, luahid_gethid(hdev));
 
 	if (data)
 		luadata_clear(data);
-
-	lua_pushinteger(L, !lua_isboolean(L, -1) ? EINVAL : 0);
+	if (lua_isboolean(L, -1)) {
+		raw->handled = lua_toboolean(L, -1);
+		raw_event_arg->ret = 0;
+	} else {
+		raw->handled = false;
```

Comment:
why? isn't it enough to `raw_event_arg->ret = !lua_isboolean(L, -1) ? EINVAL : 0`?